File: src/javascript/algorithms/tree/verticalBinaryTreeTraversal/verticalBinaryTreeTraversal.js
/**
 * Vertical Order Traversal of a Binary Tree (LeetCode 987)
 *
 * For each node at (row, col): root=(0,0), left=(row+1,col-1), right=(row+1,col+1).
 * Group by column; within same column sort by (row, value).
 *
 * Approach: BFS + Map sorted by column
 * TC: O(n log n)  SC: O(n)
 */

class TreeNode {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function verticalBinaryTreeTraversal(root) {
    if (!root) return [];

    const colMap = new Map(); // col -> [[row, val], ...]
    const queue = [[root, 0, 0]]; // [node, row, col]

    while (queue.length > 0) {
        const [node, row, col] = queue.shift();

        if (!colMap.has(col)) colMap.set(col, []);
        colMap.get(col).push([row, node.val]);

        if (node.left)  queue.push([node.left,  row + 1, col - 1]);
        if (node.right) queue.push([node.right, row + 1, col + 1]);
    }

    const sortedCols = [...colMap.keys()].sort((a, b) => a - b);
    return sortedCols.map(col => {
        const entries = colMap.get(col).sort((a, b) => a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1]);
        return entries.map(([, val]) => val);
    });
}

function build(vals) {
    if (!vals || vals.length === 0 || vals[0] === null) return null;
    const root = new TreeNode(vals[0]);
    const queue = [root];
    let i = 1;
    while (queue.length > 0 && i < vals.length) {
        const node = queue.shift();
        if (i < vals.length && vals[i] !== null) {
            node.left = new TreeNode(vals[i]);
            queue.push(node.left);
        }
        i++;
        if (i < vals.length && vals[i] !== null) {
            node.right = new TreeNode(vals[i]);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

const testCases = [
    { vals: [3, 9, 20, null, null, 15, 7], expected: [[9],[3,15],[20],[7]],       label: "[3,9,20,null,null,15,7]" },
    { vals: [1, 2, 3, 4, 5, 6, 7],         expected: [[4],[2],[1,5,6],[3],[7]],   label: "[1,2,3,4,5,6,7]" },
    { vals: [1, 2, 3, 4, 6, 5, 7],         expected: [[4],[2],[1,5,6],[3],[7]],   label: "[1,2,3,4,6,5,7] (tie-break by val)" },
    { vals: [1],                            expected: [[1]],                       label: "single node" },
    { vals: null,                           expected: [],                          label: "null" },
];

testCases.forEach(({ vals, expected, label }) => {
    const root = vals ? build(vals) : null;
    const result = verticalBinaryTreeTraversal(root);
    const status = JSON.stringify(result) === JSON.stringify(expected) ? "PASS" : "FAIL";
    console.log(`  [${status}] ${label}: ${JSON.stringify(result)} (expected ${JSON.stringify(expected)})`);
});

File: src/python/algorithms/tree/verticalBinaryTreeTraversal/vertical_binary_tree_traversal.py
"""
Vertical Order Traversal of a Binary Tree (LeetCode 987)

For each node at (row, col): root=(0,0), left=(row+1,col-1), right=(row+1,col+1).
Group by column; within same column sort by (row, value).

Approach: BFS + sorted dict
TC: O(n log n)  SC: O(n)
"""

from collections import deque, defaultdict


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def vertical_binary_tree_traversal(root: TreeNode) -> list[list[int]]:
    if not root:
        return []

    col_map = defaultdict(list)   # col -> [(row, val), ...]
    queue = deque([(root, 0, 0)]) # (node, row, col)

    while queue:
        node, row, col = queue.popleft()
        col_map[col].append((row, node.val))

        if node.left:
            queue.append((node.left, row + 1, col - 1))
        if node.right:
            queue.append((node.right, row + 1, col + 1))

    result = []
    for col in sorted(col_map):
        result.append([val for _, val in sorted(col_map[col])])
    return result


def build(vals: list) -> TreeNode | None:
    if not vals or vals[0] is None:
        return None
    root = TreeNode(vals[0])
    queue = deque([root])
    i = 1
    while queue and i < len(vals):
        node = queue.popleft()
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            queue.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            queue.append(node.right)
        i += 1
    return root


if __name__ == "__main__":
    test_cases = [
        ([3, 9, 20, None, None, 15, 7], [[9], [3, 15], [20], [7]],     "[3,9,20,null,null,15,7]"),
        ([1, 2, 3, 4, 5, 6, 7],         [[4], [2], [1, 5, 6], [3], [7]], "[1,2,3,4,5,6,7]"),
        ([1, 2, 3, 4, 6, 5, 7],         [[4], [2], [1, 5, 6], [3], [7]], "[1,2,3,4,6,5,7] (tie-break by val)"),
        ([1],                            [[1]],                           "single node"),
        (None,                           [],                              "null"),
    ]

    for vals, expected, label in test_cases:
        root = build(vals) if vals is not None else None
        result = vertical_binary_tree_traversal(root)
        status = "PASS" if result == expected else "FAIL"
        print(f"  [{status}] {label}: {result} (expected {expected})")
